src/otx/algorithms/anomaly/adapters/anomalib/callbacks/xpu.py

```diff
@@ -0,0 +1,39 @@
+"""Anomaly XPU device callback."""
+# Copyright (C) 2023 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+#
+
+import torch
+from pytorch_lightning import Callback
+
+from otx.algorithms.common.utils.utils import is_xpu_available
+
+
+class XPUCallback(Callback):
+    """XPU device callback.
+
+    Applies IPEX optimization before training, moves data to XPU.
+    """
+
+    def __init__(self, device_idx=0):
+        self.device = torch.device(f"xpu:{device_idx}")
+
+    def on_fit_start(self, trainer, pl_module):
+        """Applies IPEX optimization before training."""
+        if is_xpu_available():
+            pl_module.to(self.device)
+            model, optimizer = torch.xpu.optimize(trainer.model, optimizer=trainer.optimizers[0], dtype=torch.float32)
+            trainer.optimizers = [optimizer]
+            trainer.model = model
+
+    def on_train_batch_start(self, trainer, pl_module, batch, batch_idx):
+        """Moves train batch tensors to XPU."""
+        for k in batch:
+            if not isinstance(batch[k], list):
```

Comment:
I'm not sure if there is a case in `batch[k]` is `list`, but if so, wouldn't we consider that too?

Reply:
That happens for some of the models (detection in particular), otherwise I wouldn't have considered that case. This callback works only with anomaly models which we have right now, so this is not a general solution.